File: api/src/main/java/com/leandroadal/vortasks/entities/shop/product/GemsPackage.java
package com.leandroadal.vortasks.entities.shop.product;

import java.math.BigDecimal;
import java.math.RoundingMode;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Entity
@NoArgsConstructor
@AllArgsConstructor
public class GemsPackage {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true)
    private String nameOfPackage;
    
    private String icon;
    private Integer gems;
    private BigDecimal money;
    private Float discountPercentage;
    private int totalSales;

    public BigDecimal priceWithDiscount() {
        BigDecimal totalPrice = money;
        BigDecimal discountAmount = totalPrice.multiply(BigDecimal.valueOf(discountPercentage / 100));
        BigDecimal totalPriceWithDiscount = totalPrice.subtract(discountAmount);
        return totalPriceWithDiscount.setScale(2, RoundingMode.HALF_UP);
    }

    public void sell(){
        this.totalSales++;
    }
}

File: app/vortasks/lib/models/shop/gems_package.dart
class GemsPackage {
  final int id;
  final String name;
  final String icon;
  final int gems;
  final double money;
  final double discountPercentage;
  final double totalPrice;

  GemsPackage({
    required this.id,
    required this.name,
    required this.icon,
    required this.gems,
    required this.money,
    required this.discountPercentage,
    required this.totalPrice,
  });

  factory GemsPackage.fromJson(Map<String, dynamic> json) {
    return GemsPackage(
      id: json['id'],
      name: json['nameOfPackage'],
      icon: json['icon'],
      gems: json['gems'],
      money: json['money'],
      discountPercentage: json['discountPercentage'],
      totalPrice: json['totalPrice'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nameOfPackage': name,
      'icon': icon,
      'gems': gems,
      'money': money,
      'discountPercentage': discountPercentage,
      'totalPrice': totalPrice,
    };
  }
}
